image-registry/automate-pruning.md

```diff
@@ -0,0 +1,267 @@
+
+### Implementation Details/Notes/Constraints
+
+The implementation relies the current image pruning process invoked by the
+`oc adm prune` command. Running this in a container can be accomplished by
+using the `cli` imagestream installed by the samples operator. The image must
+be run using a service account with the `system:image-pruner` cluster role.
+
+The image registry operator will be updated to configure and reconcile all the
+components needed to run image pruning on an OpenShift cluster.
+Specifically, this includes the following:
+
+1. A `ServiceAccount` dedicated to the image pruner.
+2. A `ClusterRoleBinding` granting the `system:image-pruner` role to the pruner
+service account.
+3. A `CronJob` which runs the image pruner with the following tunable parameters:
+   1. Schedule for `CronJob`
+   2. `--keep-tag-revision` command line argument for the pruner
+   3. `--keep-younger-than` command line argument for the pruner
+   4. Resources for the `CronJob` pod template
+
+These components should be created/reconciled when the image registry operator
+is in the `Managed` state. When the image registry operator is in the `Removed`
+state, the above components should be deleted from the cluster if present.
+
+### Risks and Mitigations
+
+#### Resource consumption
+
+Risk: Image pruning requires siginficant memory to compute the blob layer graph
+needed to identify "prunable" layers.
+
+Mitigation: API will allow cluster admins to adjust the resources allocated to
+the `CronJob`.
+
+Risk: Image pruning requires too much memory to be scheduled to a node
+Risk: Image pruning exhausts resources used for application workloads
+
+Mitigation: API will allow admins to use standard tools to let pruning be run
+on desired machine pools (`affinity`, `tolerations`, `nodeSelector`)
+
+##### Concurrent pruning jobs
+
+Risk: Pruning jobs cannot run concurrently.
+Risk: Pruning job durations can exceed the interval set by the `CronJob`
+schedule.
+
+Mitigation: The `CronJob` for pruning must be configured with
+`concurrencyPolicy: Forbid`
+
+#### Removal of past pruning jobs
+
+Risk: Runs of the pruning job create another set of objects in etcd that need
+to be pruned.
+
+Mitigation: `CronJob` specifies a history limit for successful and failed jobs.
+This will be a tunable parameter via the API.
+
+#### Pruning of in-use images
+
+Risk: Pruner is not aware of a particular image, resulting in it being pruned
+prior to use.
+
+Mitigation: None at present - this is an inherent problem within the current
+pruning process.
+
+## Design Details
+
+### API
+
+Proposed changes to the image registry operator's config spec:
+
+```yaml
+spec:
+  ...
+  pruning:
+    schedule: "*/1 * * * *"
+    keepTagRevisions: 3
+    keepYoungerThan: 60m
+    resources: {}
+    affinity: {}
+    nodeSelector: {}
+    tolerations: {}
+    history:
+      successfulJobsHistoryLimit: 3
+      failedJobsHistoryLimit: 1
```

Comment:
I'm late to the party, so apologies if this came up before.

Will the config object here maintain any status conditions .... perhaps around jobs in progress?  Or history?  Or does has the CronJob status been deemed sufficient for this?

Reply:
@gabemontero good point on conditions. We haven't discussed these yet.

I'm open to suggestions, but IMO we need one to indicate if the pruner has been enabled (`ImagePrunerEnabled`), and perhaps another reporting if the most recent run was successful/failed (`ImagePruningSucceeded`)

Reply:
most recent run success/failed sounds good;

were you yay/nay on an "job in progress"? .... depends on how long prune jobs go for I suppose...

In theory, each config value under spec should have a corresponding value under status, especially if there are reasons a config setting would not be applied (like a job is in progress, and the job takes a long time).